packages/pigeon/lib/kotlin/templates.dart

```diff
@@ -0,0 +1,256 @@
+ *
+ * <p>Objects stored in this container are represented by an object in Dart that is also stored in
+ * an InstanceManager with the same identifier.
+ *
+ * <p>When an instance is added with an identifier, either can be used to retrieve the other.
+ *
+ * <p>Added instances are added as a weak reference and a strong reference. When the strong
+ * reference is removed with [remove] and the weak reference is deallocated, the
+ * `finalizationListener` is made with the instance's identifier. However, if the strong reference
+ * is removed and then the identifier is retrieved with the intention to pass the identifier to Dart
+ * (e.g. calling [getIdentifierForStrongReference]), the strong reference to the
+ * instance is recreated. The strong reference will then need to be removed manually again.
+ */
+@Suppress("UNCHECKED_CAST", "MemberVisibilityCanBePrivate", "unused")
+class $instanceManagerClassName(private val finalizationListener: $_finalizationListenerClassName) {
+  /** Interface for listening when a weak reference of an instance is removed from the manager.  */
+  interface $_finalizationListenerClassName {
+    fun onFinalize(identifier: Long)
+  }
+
+  private val identifiers = java.util.WeakHashMap<Any, Long>()
+  private val weakInstances = HashMap<Long, java.lang.ref.WeakReference<Any>>()
+  private val strongInstances = HashMap<Long, Any>()
+  private val referenceQueue = java.lang.ref.ReferenceQueue<Any>()
+  private val weakReferencesToIdentifiers = HashMap<java.lang.ref.WeakReference<Any>, Long>()
+  private val handler = android.os.Handler(android.os.Looper.getMainLooper())
+  private var nextIdentifier: Long = minHostCreatedIdentifier
+  private var hasFinalizationListenerStopped = false
+
+  /**
+   * Modifies the time interval used to define how often this instance removes garbage collected
+   * weak references to native Android objects that this instance was managing.
+   */
+  var clearFinalizedWeakReferencesInterval: Long = 3000
+    set(value) {
+      handler.removeCallbacks { this.releaseAllFinalizedInstances() }
+      field = value
+      releaseAllFinalizedInstances()
+    }
+
+  init {
+    handler.postDelayed(
+      { releaseAllFinalizedInstances() },
+      clearFinalizedWeakReferencesInterval
+    )
+  }
+
+  companion object {
+    // Identifiers are locked to a specific range to avoid collisions with objects
+    // created simultaneously from Dart.
+    // Host uses identifiers >= 2^16 and Dart is expected to use values n where,
+    // 0 <= n < 2^16.
+    private const val minHostCreatedIdentifier: Long = 65536
+    private const val tag = "$instanceManagerClassName"
+
+    /**
+     * Instantiate a new manager.
+     *
+     *
+     * When the manager is no longer needed, [stopFinalizationListener] must be called.
+     *
+     * @param finalizationListener the listener for garbage collected weak references.
+     * @return a new `$instanceManagerClassName`.
```

Comment:
Optional (for all these comments): the Kotlin style guide [recommends against Javadoc style](https://kotlinlang.org/docs/coding-conventions.html#documentation-comments).

Reply:
Ideally we would just use the addDocumentationComments method for this, but idk if it's too complicated to plumb into these templates.

Reply:
It would be pretty complicated since `addDocumentationComments` requires an `Indent` and the helper method returns `<String>[]`.